numpy/core/src/multiarray/methods.c

```diff
@@ -2694,6 +2696,161 @@ array_complex(PyArrayObject *self, PyObject *NPY_UNUSED(args))
+
+    if (PyDataType_ISSIGNED(dtype)) {
+        managed_dtype.code = kDLInt;
+    } else if (PyDataType_ISUNSIGNED(dtype)) {
+        managed_dtype.code = kDLUInt;
+    } else if (PyDataType_ISFLOAT(dtype)) {
+        // We can't be sure that the dtype is
+        // IEEE or padded.
+        if (itemsize > 8) {
+            PyErr_SetString(PyExc_TypeError, "DLPack only supports IEEE "
+                    "floating point types without padding.");
+            return NULL;
+        } 
+        managed_dtype.code = kDLFloat;
+    } else if (PyDataType_ISCOMPLEX(dtype)) {
+        // We can't be sure that the dtype is
+        // IEEE or padded.
+        if (itemsize > 16) {
+            PyErr_SetString(PyExc_TypeError, "DLPack only supports IEEE "
+                    "complex point types without padding.");
+            return NULL;
+        } 
+        managed_dtype.code = kDLComplex;
+    } else {
+        PyErr_SetString(PyExc_TypeError,
+                        "DLPack only supports signed/unsigned integers, float "
+                        "and complex dtypes.");
+        return NULL;
+    }
+
+    DLManagedTensor *managed = PyMem_Malloc(sizeof(DLManagedTensor));
+    if (managed == NULL) {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate the DLManagedTensor struct.");
+        return NULL;
+    }
+
+    managed->dl_tensor.data = PyArray_DATA(self);
+    managed->dl_tensor.device.device_type = kDLCPU;
+    managed->dl_tensor.device.device_id = 0;
+    managed->dl_tensor.dtype = managed_dtype;
+
+
+    int64_t *managed_shape_strides = PyMem_Malloc(sizeof(int64_t) * ndim * 2);
+    if (managed_shape_strides == NULL) {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate the DLManagedTensor struct shape/"
+                        "strides.");
+        PyMem_Free(managed);
+        return NULL;
+    }
+
+    int64_t *managed_shape = managed_shape_strides;
+    int64_t *managed_strides = managed_shape_strides + ndim;
+    for (int i = 0; i < ndim; ++i) {
+        managed_shape[i] = shape[i];
+        // Strides in DLPack are items; in NumPy are bytes.
+        managed_strides[i] = strides[i] / itemsize;
+    }
+
+    managed->dl_tensor.ndim = ndim;
+    managed->dl_tensor.shape = managed_shape;
+    managed->dl_tensor.strides = managed_strides;
+    managed->dl_tensor.byte_offset = 0;
+    managed->manager_ctx = self;
+    managed->deleter = array_dlpack_deleter;
+
+    PyObject *capsule = PyCapsule_New(managed, NPY_DLPACK_CAPSULE_NAME,
+            array_dlpack_capsule_deleter);
+    if (capsule == NULL) {
+        PyMem_Free(managed);
+        PyMem_Free(managed_shape_strides);
+        return NULL;
+    }
+    
+    // the capsule holds a reference
+    PyArray_INCREF(self);
+    return capsule;
+}
+
+static PyObject *
+array_dlpack_device(PyArrayObject *NPY_UNUSED(self), PyObject *NPY_UNUSED(args))
+{
+    return Py_BuildValue("ii", kDLCPU, 0);
```

Comment:
Note: This could be `kDLCPUPinned` if [the NumPy array is backed by pinned memory](https://docs.cupy.dev/en/stable/reference/generated/cupyx.empty_pinned.html#cupyx.empty_pinned), but let's get this in first, as we need a mechanism to retrieve this information (maybe #17582?).

Reply:
How do you reliably free such memory when the `np.ndarray` is destroyed? This may be useful when I build `from_dlpack`.

Reply:
We don't have to free the memory, the owner is in charge of that (with the deleter)?  I.e. all you do is hold on to the "capsule" as the base object (or at least its `deleter`).

There might be a small issue, that NumPy exposes the `base` object.  DLPack has a "there is only one consumer" contract, but the easiest way would be for NumPy to store the capsule as `arr.base`.  This capsule will *not* be renamed, since it still _owns_ the data (must call the `deleter`).
But such a capsule is not strictly guarded against being consumed a second time.  And that would be wrong, since it would steal ownership from the NumPy array.

Enforcing that would mean wrapping the dlpack capsule into a new object for `arr.base`.  If we need to wrap it into a new object, maybe @eric-wieser's point about a `dlpack` module (possibly even going through the buffer protocol as the final step), becomes more attractive.
(To be clear: It doesn't bother me right now, we could probably transition easy enough.)

About `kDLCPUPinned`: It would be nice if DLPack had an easy way to tell that memory is transparently accessible from the CPU if there are different devices that are.  (A flag would be cute for compatibility if new devices are added, but maybe just a macro?)

Reply:
How about the following: We rename the existing capsule to `"used_dltensor"`, create a new capsule containing the DLPack struct and a new name. This way, the old capsule was consumed and the contract isn't violated, and we're just holding onto the reference of a NumPy "internal" DLPack capsule.

Reply:
Yeah, should work well, any intermediate object seems fine.  Or maybe I am misunderstanding and the renamed capsule is always still the owner and must always only be deleted when the data is not needed anymore.  (I.e. in this PR the rename should not prevent cleanup.)  In that case we can just rename and attach the capsule, with the only downside being that `.base` doesn't provide much useful info.